src/messaging/Channel.h

```diff
@@ -0,0 +1,195 @@
+ *    ChannelBuilder can be safely released.
+ */
+class ChannelBuilder
+{
+public:
+    enum class NetworkPreference
+    {
+        kNetwork_Default,
+        kNetwork_HighBandwidth,
+        kNetwork_LowLatency,
+        kNetwork_LowPowerConsumption,
+    };
+
+    enum class TransportPreference
+    {
+        kTransport_Connectionless,           // default
+        kTransport_PreferConnectionOriented, // will fallback to connectionless TCP is not supported
+        kTransport_ConnectionOriented,       // will fail if TCP is not supported
+    };
+
+    enum class SessionType
+    {
+        kSession_NoEncryption,
+        kSession_PASE, // Use SPAKE2 key exchange
+        kSession_CASE, // Use SIGMA key exchange
+    };
+
+    ChannelBuilder & SetPeerNodeId(NodeId peerNodeId)
+    {
+        mPeerNodeId = peerNodeId;
+        return *this;
+    }
+    NodeId GetPeerNodeId() { return mPeerNodeId; }
+
+    ChannelBuilder & SetNetworkPreference(NetworkPreference preference)
+    {
+        mNetworkPreference = preference;
+        return *this;
+    }
+    NetworkPreference GetNetworkPreference() { return mNetworkPreference; }
+
+    ChannelBuilder & SetTransportPreference(TransportPreference preference)
+    {
+        mTransportPreference = preference;
+        return *this;
+    }
+    TransportPreference GetTransportPreference() { return mTransportPreference; }
+
+    ChannelBuilder & SetSessionType(SessionType preference)
+    {
+        mSessionType = preference;
+        return *this;
+    }
+    SessionType GetSessionType() { return mSessionType; }
+
+private:
+    NodeId mPeerNodeId;
+    NetworkPreference mNetworkPreference     = NetworkPreference::kNetwork_Default;
+    TransportPreference mTransportPreference = TransportPreference::kTransport_Connectionless;
+    SessionType mSessionType                 = SessionType::kSession_CASE;
+};
+
+enum class ChannelState
+{
+    kChanneState_None,
+    kChanneState_Preparing,
+    kChanneState_Ready,
+    kChanneState_Closed,
+    kChanneState_Failed,
+};
+
+class ChannelContext;
+
+/**
+ *  @brief
+ *    ChannelHandle is a reference to a channel. An active ChannelHandle will
+ *    keep the channel available and ready for use, such that a message can be
+ *    sent immediately to the peer.
+ *
+ *    The ChannelHandle controls the lifespan of the channel. The ChannelHandle
+ *    objects are not copyable, so there can be only one instance referring to
+ *    the channel. When the handler is released, the channel will be flagged as
+ *    pending close, and if there is no active exchange which is using the
+ *    channel, the channel will be closed.
+ *
+ *    The ChannelHandle will track channel status, and notify applications
+ *    when the channel state changes via ChannelDelegate.
+ */
+class ChannelHandle
+{
+public:
+    ChannelHandle(ChannelContext * channelContext) : mChannelContext(channelContext) {}
+    ~ChannelHandle() { Release(); } // Release on destruction
```

Comment:
remove comment (restates the code)

Is there an Aquire? Otherwise why is Release() public?

Reply:
I'll remove the comments.

This is not reference counted, it works like unique ptr, its ownership initialized during construction.